### PR TITLE
EXP-23 Edit expense + validations and add expense touchups

### PR DESCRIPTION
**Pull request by zsoltibv**

### CODE REVIEW

File: src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java [hunk 1]
```diff
@@ -1,46 +1,97 @@
...
+        } else if (!currencyRepository.existsById(expenseDto.getCurrencyId())) {
+            throw new NotFoundException("currencyId", "Currency with ID " + expenseDto.getCurrencyId() + " not found");
+        }
+
+        expenseRepository.save(ExpenseMapper.toExpense(expenseDto));
+    }
+
+    @Override
+    public void editExpense(Integer expenseId, ExpenseDto expenseDto) throws NotFoundException {
+        Optional<Expense> existingExpense = expenseRepository.findById(expenseId);
```
Comment (FitaRoberttt) [inline src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java:54]: I think you can simplify here by using `expenseRepository.findById(expenseId).orElseThrow()`. This eliminates the need for the `if` condition and `Optional`.

**Review by FitaRoberttt — APPROVED** [review]

**Review by en-dragos-dordea — CHANGES_REQUESTED** [review]

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 2]
```diff
@@ -17,18 +21,25 @@
...
     private final ExpenseService expenseService;
 
-    public ExpenseController(ExpenseService expenseService){
+    public ExpenseController(ExpenseService expenseService) {
         this.expenseService = expenseService;
     }
 
     @PostMapping()
-    public ResponseEntity<ExpenseDto> addExpense(@RequestBody @Valid ExpenseDto expenseDto){
+    public ResponseEntity<ExpenseDto> addExpense(@RequestBody @Valid ExpenseDto expenseDto) throws RuntimeException {
```
Comment (en-dragos-dordea) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:29]: No need to specify that the method throws `RuntimeException`, as this kind of exception is unchecked. Please remove the throws from methods signatures

File: src/main/java/com/endava/expensesmanager/exception/handler/ApiExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public ResponseEntity<ApiErrorSingle> handleCurrencyNotFoundException(CurrencyNotFoundException e,
+                                                                          ServletWebRequest request) {
+        return getApiErrorResponseEntity(request, e.getMessage(), HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(UserNotFoundException.class)
+    public ResponseEntity<ApiErrorSingle> handleUserNotFoundException(UserNotFoundException e,
+                                                                          ServletWebRequest request) {
+        return getApiErrorResponseEntity(request, e.getMessage(), HttpStatus.BAD_REQUEST);
+    }
```
Comment (en-dragos-dordea) [inline src/main/java/com/endava/expensesmanager/exception/handler/ApiExceptionHandler.java:63]: These exceptions can be handled in the same method - `@ExceptionHandler` can take an array

File: src/main/java/com/endava/expensesmanager/model/mapper/ExpenseMapper.java [hunk 4]
```diff
@@ -43,4 +51,18 @@ public static ExpenseDto toDto(Expense expense)
 
         return expenseDto;
     }
+
+    public static Expense toUpdatedExpense(Expense existingExpense, ExpenseDto expenseDto, User user, Category category, Currency currency)
+            throws ExpenseNotFoundException, UserNotFoundException, CategoryNotFoundException, CurrencyNotFoundException {
```
Comment (en-dragos-dordea) [inline src/main/java/com/endava/expensesmanager/model/mapper/ExpenseMapper.java:56]: The throws can be removed from here too

**Review by en-dragos-dordea — CHANGES_REQUESTED** [review]

**Review by en-dragos-dordea — APPROVED** [review]

**Pull request merged**